httpbp/client.go

```diff
@@ -0,0 +1,164 @@
+		}
+		return r, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return rsp.(*http.Response), nil
+}
+
+// CircuitBreaker enables circuit breaking for your client.
+func CircuitBreaker(minRequests uint32, minFailureRatio float64) LowLevelClientMiddleware {
+	settings := gobreaker.Settings{
+		Interval: 60 * time.Second, // Reset the counts every 60 seconds
+		ReadyToTrip: func(counts gobreaker.Counts) bool {
+			failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+			return counts.Requests >= minRequests && failureRatio >= minFailureRatio
+		},
+	}
+	breaker := gobreaker.NewCircuitBreaker(settings)
+
+	return func(name string, next http.RoundTripper) http.RoundTripper {
+		return &circuitBreaker{
+			next:    next,
+			breaker: breaker,
+		}
+	}
+}
+
+type spanWrapper struct {
+	name string
+	next http.RoundTripper
+}
+
+func (s *spanWrapper) RoundTrip(req *http.Request) (*http.Response, error) {
+	_, ctx := opentracing.StartSpanFromContext(
+		req.Context(),
+		s.name,
+		tracing.SpanTypeOption{Type: tracing.SpanTypeClient},
+	)
+	rsp, err := s.next.RoundTrip(req)
+	if span := opentracing.SpanFromContext(ctx); span != nil {
+		span.FinishWithOptions(tracing.FinishOptions{
+			Ctx: req.Context(),
+			Err: err,
+		}.Convert())
+	}
+	return rsp, err
+}
+
+// SpanWrapper enables tracing for your client
+func SpanWrapper(suffix string) LowLevelClientMiddleware {
+	return func(name string, next http.RoundTripper) http.RoundTripper {
+		return &spanWrapper{
+			name: name + suffix,
+			next: next,
+		}
+	}
+}
+
+// ClientConfig provides options for configuring a typical http client.
+type ClientConfig struct {
+	MaxConnsPerHost int
+	Timeout         time.Duration
+	Middlewares     []LowLevelClientMiddleware
+}
+
+// NewTypicalClientConfig contains the settings that feel reasonable
+// under many conditions.  If you are calling out to a very
+// high-performance and/or concurrent service, you may want to
+// adjust values to maximize your performance. If you are calling
+// out to a low-performance service, these values should be better
+// than defaults, but you may still want to adjust in order to
+// maximize safety.
+func NewTypicalClientConfig() *ClientConfig {
+	return &ClientConfig{
+		MaxConnsPerHost: 20,
+		Timeout:         1 * time.Second,
+		Middlewares: []LowLevelClientMiddleware{
+			CircuitBreaker(100, 0.1),
+			MaxConcurrency(50),
+			SpanWrapper(""),
+		},
+	}
+}
+
+// NewClient will create an HttpClient with the configured transport and
+// transport middleware.  Note that retries are not suitable for low-level
+// middleware, and you should consider bringing your own retries, wrapping
+// this client.
+func NewClient(config *ClientConfig) *http.Client {
+	var tripper http.RoundTripper = &http.Transport{
+		MaxConnsPerHost: config.MaxConnsPerHost,
+	}
+	for i := len(config.Middlewares) - 1; i >= 0; i-- {
+		mw := config.Middlewares[i]
+		tripper = mw(fmt.Sprintf("%v", mw), tripper)
```

Comment:
What is the intention of using `fmt.Sprintf("%v", mw)`? I don't see you define stringer in any of the middleware implementations, so this would look like some `"{foo <nil>}"` (see https://play.golang.org/p/EkRg7IkWS4b), and that's not suitable to be used in a span name.